libgadget/metal_return.c

```diff
@@ -1021,3 +1021,435 @@ stellar_density(const ActiveParticles * act, MyFloat * StellarAges, MyFloat * Ma
+        /* Discard if no room, otherwise keep*/
+        if(listsize < NUMNB) {
+            list[listsize].part = part;
+            list[listsize].dist2 = dist2;
+        }
+    }
+    /* Bracket the interval*/
+    else if(list[0].dist2 >= dist2) {
+        memmove(list+1, list, (newsize-1) * sizeof(list[0]));
+        list[0].part = part;
+        list[0].dist2 = dist2;
+    }
+    /* Do a bisection: now the new particle belongs somewhere in the list.*/
+    else {
+        int max = listsize-1, min=0;
+        while(max - min > 1)
+        {
+            int mid = (max + min) / 2;
+            if(list[mid].dist2 > dist2)
+                max = mid;
+            else
+                min = mid;
+        }
+        memmove(list+max+1, list+max, (newsize-max-1) * sizeof(list[0]));
+        list[max].part = part;
+        list[max].dist2 = dist2;
+    }
+    return newsize;
+}
+/* Find an initial k neighbours set. This searches for gas particles
+ * from the current index of the star particle, which should be ok as particles are in Peano-Hilbert order.
+ * Note: if there are < NUMNB gas particles on the current processor this will not fill the queue.*/
+static void
+stellar_knn_find_topnode(int place, TreeWalk * tw)
+{
+    int pi = P[place].PI;
+    int topleaf = domain_get_topleaf(P[place].Key, STELLAR_KNN_GET_PRIV(tw)->ddecomp);
+    STELLAR_KNN_GET_PRIV(tw)->topnode[pi] = tw->tree->TopLeaves[topleaf].treenode;
+    STELLAR_KNN_GET_PRIV(tw)->nnb[pi] = 0;
+    return;
+}
+
+static void
+stellar_knn_copy(int place, TreeWalkQueryStellarKNN * I, TreeWalk * tw)
+{
+    int pi = P[place].PI;
+    memmove(I->neighbours, &STELLAR_KNN_GET_PRIV(tw)->Neighbours[NUMNB * pi], sizeof(STELLAR_KNN_GET_PRIV(tw)->Neighbours[0])* NUMNB);
+    I->nnb = STELLAR_KNN_GET_PRIV(tw)->nnb[pi];
+    I->topnode = STELLAR_KNN_GET_PRIV(tw)->topnode[pi];
+}
+
+static void
+stellar_knn_reduce(int place, TreeWalkResultStellarKNN * remote, enum TreeWalkReduceMode mode, TreeWalk * tw)
+{
+    /* We need to merge the lists if mode == 1, otherwise we just copy it.*/
+    int pi = P[place].PI;
+    struct Neighbour * neigh = &STELLAR_KNN_GET_PRIV(tw)->Neighbours[NUMNB * pi];
+    if(mode == 0) {
+        memmove(neigh, remote->neighbours, sizeof(STELLAR_KNN_GET_PRIV(tw)->Neighbours[0])* NUMNB);
+        STELLAR_KNN_GET_PRIV(tw)->nnb[pi] = remote->nnb;
+    }
+    else {
+        /* Walk through the new list until we find someone not in the original list.
+         * Add them, then continue.*/
+        int i;
+        for(i = 0; i < remote->nnb; i++) {
+            /* Check whether the list changed: if the furthest particle is still the same we have no merging to do.*/
+            if(remote->nnb == STELLAR_KNN_GET_PRIV(tw)->nnb[pi]
+                    && remote->neighbours[remote->nnb-1].part == neigh[remote->nnb-1].part)
+                break;
+            const struct Neighbour nn = remote->neighbours[i];
+            if(nn.part == neigh[i].part)
```

Comment:
I think this comparison requires part to be globally unique (e.g. particle ID).

Reply:
Indeed! Well noticed, Yu!